### PR TITLE
Initial implementation of multiplication operator for MKLDNN

### PR DESCRIPTION
**Pull request by mozga-intel**

### CODE REVIEW

File: python/paddle/fluid/layers/nn.py [hunk 1]
```diff
@@ -159,67 +159,37 @@ def fc(input,
...
+            inputs={"X": input_var,
+                    "Y": w},
             outputs={"Out": tmp},
             attrs={
-                "use_mkldnn": use_mkldnn,
-                "is_test": is_test,
-                "bias_attr": bias_attr
+                "x_num_col_dims": num_flatten_dims,
+                "y_num_col_dims": 1,
+                "use_mkldnn": use_mkldnn
```
Comment (luotao1) [inline python/paddle/fluid/layers/nn.py]: why use `mul_mkldnn_op` to replace `fc_mkldnn_op` here?
In my mind, the performance of `fc_mkldnn_op` is better than `mul_mkldnn_op` + `elementwise_add_op`.

File: paddle/fluid/platform/mkldnn_helper.h [hunk 2]
```diff
@@ -34,6 +33,32 @@ typedef std::unique_ptr<MKLDNNMemory> MKLDNNMemoryPtr;
...
+  return std::shared_ptr<tf_pd<Type>>(pd);
+}
+
+template <typename Type, typename Engine, typename Primitive, typename... Args>
+tf_pd<Type> MKLDNNBwdPrimitiveDesc(const Engine& e, const Primitive& p,
+                                   Args&&... args) {
+  auto desc = tf_desc<Type>(args...);
+  return tf_pd<Type>(desc, e, p);
+}
+
```
Comment (luotao1) [inline paddle/fluid/platform/mkldnn_helper.h]: Are line 36-61 only used in `mul_mkldnn_op`? If these lines are not used in other ops, they should be moved into `mul_mkldnn_op` internel.

Comment (mozga-intel) [reply paddle/fluid/platform/mkldnn_helper.h]: At this moment, yes, but this code may be uses by the all mkldnn operators.

File: python/paddle/fluid/layers/nn.py (hunk 1 above)
Comment (mozga-intel) [reply python/paddle/fluid/layers/nn.py]: Yes I must agree with you, but the fully connected layer for mkldnn has a little problem with the input of matrices, What means that in the case when the input of layer is greater than one the mkldnn operator doesn't work correctly. Therefore we hit upon a good idea to remove this pitfall and withdraw the fully connected operator. Moreover this simple solution give us opportunity to run the new neural network models.

Comment (mozga-intel) [reply python/paddle/fluid/layers/nn.py]: With a heavy heart, I must say, yes, this operator is removed but for a while. I guess that in the future we will come back to the fully connected layer for a mkldnn.

**Comment by luotao1** [comment]
@tensor-tang Do you have any other comments about this PR?

File: paddle/fluid/platform/mkldnn_helper.h [hunk 3]
```diff
@@ -34,6 +33,32 @@ typedef std::unique_ptr<MKLDNNMemory> MKLDNNMemoryPtr;
...
+template <typename Type>
+void* to_void_cast(const Type* t) {
+  return static_cast<void*>(const_cast<Type*>(t));
+}
+
+template <class Type>
+using tf_desc = typename Type::desc;
+
+template <class Type>
+using tf_pd = typename Type::primitive_desc;
```
Comment (tensor-tang) [inline paddle/fluid/platform/mkldnn_helper.h]: A small question, why named `tf_pd` here?

**Pull request merged**